move flushing of uncommitted ptree nodes out of EncodeValue

rafael-atticlabs opened the pull request:
Ok, here's step one. 

Separate the concern of flushing uncommitted (linked in-memory) ptree nodes from encoding.

Towards #3525

cmasone-attic commented on go/types/meta_sequence.go:
```diff
@@ -339,3 +339,31 @@ func (es emptySequence) getCompositeChildSequence(start uint64, length uint64, h
+
+	v.WalkValues(func(v Value) {
+		iterateUncommittedChildren(v, cb)
+	})
```
It's surprising that you do this here at the end. The comment, and function name, lead me to believe this function is all about visiting uncommitted children in prolly trees, which you explicitly handle above. What's this here for? I'm sure it's to cover some necessary case, (maybe explicit Refs in collections or structs...?) but an example in a comment would help me understand.

rafael-atticlabs replied:
Yeah. I didn't at first and it didn't work. The issue is that You can nest collections that have uncommitted values within other values (i.e. a struct) and you need to find those. So you need to recursively walk the value, but only "leave" the current chunk when traversing an uncommitted ref. I'll add a comment.

cmasone-attic commented on go/types/opcache.go:
```diff
@@ -335,10 +335,10 @@ func decodeValue(bs []byte, asValue bool, vr ValueReader) ([]byte, Value) {
 }
 
 // Note that, if 'v' are prolly trees, any in-memory child chunks will be written to vw at this time.
```
Is this comment still true? I'm not entirely sure how it could be, and that would be a problem, no?

rafael-atticlabs replied:
good catch. i think that probably does need to flush uncommitted child nodes as well.

rafael-atticlabs commented:
Ok. PTAL. I basically needed to undo edits to a few files here, so I flattened the commits again. The only changes since the last commit are in opcache and comments on iterateUncommittedChildren

cmasone-attic commented:
lgtm
